File: scripts/simple-scene.py
#!/usr/local/bin/yade-trunk -x
# -*- encoding=utf-8 -*-

## Omega is the super-class that orchestrates the whole program.
## It holds the entire simulation (MetaBody), takes care of loading/saving,
## starting/stopping the simulation, loads plugins and more.

o=Omega() # for advaned folks: this creates default MetaBody as well

## Initializers are run before the simulation.
o.initializers=[
	## Create bounding boxes. They are needed to zoom the 3d view properly before we start the simulation.
	BoundDispatcher([InteractingSphere2AABB(),InteractingBox2AABB(),])
]

## Engines are called consecutively at each iteration. Their order matters.
##
## Some of them work by themselves (GlobalEngine, PartialEngine - the difference of these two is unimportant).
##
## MetaEngines act as dispatchers and based on the type of objects they operate on, different EngineUnits are called.
o.engines=[
	## Resets forces and momenta the act on bodies
	BexResetter(),
	## associates bounding volume - in this case, AxisAlignedBoundingBox (AABB) - to each body.
	## MetaEngine calls corresponding EngineUnit, depending on whether the body is Sphere, BoxModel, or MetaBody (rootBody).
	## AABBs will be used to detect collisions later, by PersistentSAPCollider
	BoundDispatcher([
		InteractingSphere2AABB(),
		InteractingBox2AABB(),
	]),
	## Using bounding boxes created by the previous engine, find possible body collisions.
	## These possible collisions are inserted in Omega.interactions container (MetaBody::transientInteractions in c++).
	InsertionSortCollider(),
	## Decide whether the potential collisions are real; if so, create geometry information about each potential collision.
	## Here, the decision about which EngineUnit to use depends on types of _both_ bodies.
	## Note that there is no EngineUnit for box-box collision. They are not implemented.
	InteractionGeometryDispatcher([
		Ig2_Sphere_Sphere_ScGeom(),
		Ig2_Box_Sphere_ScGeom()
	]),
	## Create physical information about the interaction.
	## This may consist in deriving contact rigidity from elastic moduli of each body, for example.
	## The purpose is that the contact may be "solved" without reference to related bodies,
	## only with the information contained in contact geometry and physics.
	InteractionPhysicsDispatcher([SimpleElasticRelationships()]),
	## "Solver" of the contact, also called (consitutive) law.
	## Based on the information in interaction physics and geometry, it applies corresponding forces on bodies in interaction.
	ElasticContactLaw(),
	## Apply gravity: all bodies will have gravity applied on them.
	## Note the engine parameter 'gravity', a vector that gives the acceleration.
	GravityEngine(gravity=[0,0,-9.81]),
	## Forces acting on bodies are damped to artificially increase energy dissipation in simulation.
	## (In this model, the restitution coefficient of interaction is 1, which is not realistic.)
	## This MetaEngine acts on all PhysicalActions and selects the right EngineUnit base on type of the PhysicalAction.
	#
	# note that following 4 engines (till the end) can be replaced by an optimized monolithic version:
	NewtonIntegrator(damping=0.1)
	#
#	PhysicalActionDamper([
#		CundallNonViscousForceDamping(damping=0.2),
#		CundallNonViscousMomentumDamping(damping=0.2)
#	]),
#	## Now we have forces and momenta acting on bodies. Newton's law calculates acceleration that corresponds to them.
#	PhysicalActionApplier([
#		NewtonsForceLaw(),
#		NewtonsMomentumLaw(),
#	]),
#	## Acceleration results in velocity change. Integrating the velocity over dt, position of the body will change.
#	StateMetaEngine([LeapFrogPositionIntegrator()]),
#	## Angular acceleration changes angular velocity, resulting in position and/or orientation change of the body.
#	StateMetaEngine([LeapFrogOrientationIntegrator()])
]


## The yade.utils module contains some handy functions, like yade.utils.box and yade.utils.sphere.
## After this import, they will be accessible as utils.box and utils.sphere.
from yade import utils

## create bodies in the simulation: one box in the origin and one floating above it.
##
## The box:
## * extents: half-size of the box. [.5,.5,.5] is unit cube
## * center: position of the center of the box
## * dynamic: it is not dynamic, i.e. will not move during simulation, even if forces are applied to it
## * color: for the 3d display; specified  within unit cube in the RGB space; [1,0,0] is, therefore, red
## * young: Young's modulus
## * poisson: Poissons's ratio

o.bodies.append(utils.box(center=[0,0,0],extents=[.5,.5,.5],color=[0,0,1],dynamic=False))

## The above command could be actully written without the util.box function like this:
## (will not be executed, since the condition is never True)
if False:
	# Create empty body object
	b=Body()
	# set the isDynamic body attribute
	b['isDynamic']=False
	# Assign geometrical model (shape) to the body: a box of given size
	b.shape=BoxModel(extents=[.5,.5,.5],diffuseColor=[1,0,0])
	# Assign computational model (mold; may be simplified form of shape) to the body
	b.mold=InteractingBox(extents=[.5,.5,.5],diffuseColor=[1,0,0])
	# physical parameters:
	# store mass to a temporary
	mass=8*.5*.5*.5*2400
	# * se3 (position & orientation) as 3 position coordinates, then 3 direction axis coordinates and rotation angle
	b.phys=BodyMacroParameters(se3=[0,0,0,1,0,0,0],mass=mass,inertia=[mass*4*(.5**2+.5**2),mass*4*(.5**2+.5**2),mass*4*(.5**2+.5**2)],young=30e9,poisson=.3)
	# other information about AABB will be updated during simulation by relevant BoundDispatcher
	b.bound=AABB(diffuseColor=[0,1,0])
	# add the body to the simulation
	o.bodies.append(b)


## The sphere
##
## * First two arguments are radius and center, respectively. They are used as "positional arguments" here:
## python will deduce based on where they are what they mean.
##
## It could also be written without using utils.sphere - see gui/py/utils.py for the code of the utils.sphere function
o.bodies.append(utils.sphere([0,0,2],1,color=[0,1,0]))

## Estimate timestep from p-wave speed and multiply it by safety factor of .2
o.dt=.2*utils.PWaveTimeStep()

## Save the scene to file, so that it can be loaded later. Supported extension are: .xml, .xml.gz, .xml.bz2.
o.save('/tmp/a.xml.bz2');
#o.run(100000); o.wait(); print o.iter/o.realtime,'iterations/sec'

def onBodySelect(id):
	print "Selected:",id
	utils.highlightNone()
	for i in O.interactions.withBody(id):
		O.bodies[i.id2 if i.id1==id else i.id1].shape['highlight']=True
		print i.id1,i.id2,i.phys,i.geom

from yade import qt
qt.Controller()

File: attic/pkg/dem/PreProcessor/SimpleScene.cpp
#include"SimpleScene.hpp"
#include<yade/pkg-dem/Shop.hpp>
#include<yade/pkg-dem/BodyMacroParameters.hpp>
#include<yade/pkg-common/Bo1_Sphere_Aabb.hpp>
#include<yade/pkg-common/Bo1_Box_Aabb.hpp>
#include<yade/pkg-common/ForceResetter.hpp>
#include<yade/pkg-common/StateMetaEngine.hpp>
#include<yade/pkg-common/Aabb.hpp>
#include<yade/pkg-common/Box.hpp>
#include<yade/pkg-common/NewtonsForceLaw.hpp>
#include<yade/pkg-common/NewtonsMomentumLaw.hpp>
#include<yade/pkg-common/LeapFrogPositionIntegrator.hpp>
#include<yade/pkg-common/LeapFrogOrientationIntegrator.hpp>
#include<yade/pkg-dem/Ig2_Sphere_Sphere_ScGeom.hpp>
#include<yade/pkg-dem/Ig2_Box_Sphere_ScGeom.hpp>
#include<yade/pkg-common/StateMetaEngine.hpp>
#include<yade/pkg-common/InteractionGeometryDispatcher.hpp>
#include<yade/pkg-common/InteractionPhysicsDispatcher.hpp>
#include<yade/pkg-common/BoundDispatcher.hpp>
#include<yade/pkg-common/PhysicalActionDamper.hpp>
#include<yade/pkg-common/PhysicalActionApplier.hpp>
#include<yade/pkg-common/CundallNonViscousDamping.hpp>
#include<yade/pkg-common/CundallNonViscousDamping.hpp>
#include<yade/pkg-common/GravityEngines.hpp>
#include<yade/pkg-dem/ElasticContactLaw.hpp>
#include<yade/pkg-dem/ScGeom.hpp>
#include<yade/pkg-dem/Ip2_FrictMat_FrictMat_FrictPhys.hpp>
#include<yade/pkg-common/InsertionSortCollider.hpp>



YADE_PLUGIN((SimpleScene));
bool SimpleScene::generate(){
	message="";
	//@
	rootBody=Shop::rootBody();
	//@
	/* initializers */
		rootBody->initializers.clear();
		//@
		shared_ptr<BoundDispatcher> boundDispatcher	= shared_ptr<BoundDispatcher>(new BoundDispatcher);
			boundDispatcher->add(new Bo1_Sphere_Aabb);
			boundDispatcher->add(new Bo1_Box_Aabb);
			rootBody->initializers.push_back(boundDispatcher);
	//@
	/* engines */
		rootBody->engines.clear();
		//@
		rootBody->engines.push_back(shared_ptr<Engine>(new ForceResetter));
		//@
		// use boundDispatcher that we defined above
		rootBody->engines.push_back(boundDispatcher);
		//@
		shared_ptr<InsertionSortCollider> collider(new InsertionSortCollider);
			rootBody->engines.push_back(collider);
		//@
		shared_ptr<InteractionGeometryDispatcher> igeomDispatcher(new InteractionGeometryDispatcher);
			igeomDispatcher->add(new Ig2_Sphere_Sphere_ScGeom);
			igeomDispatcher->add(new Ig2_Box_Sphere_ScGeom);
			rootBody->engines.push_back(igeomDispatcher);
		//@
		shared_ptr<InteractionPhysicsDispatcher> iphysDispatcher(new InteractionPhysicsDispatcher);
			iphysDispatcher->add(new Ip2_FrictMat_FrictMat_FrictPhys);
			rootBody->engines.push_back(iphysDispatcher);
		//@
		shared_ptr<ElasticContactLaw> ecl(new ElasticContactLaw);
			rootBody->engines.push_back(ecl);
		//@
		shared_ptr<GravityEngine> ge(new GravityEngine);
			ge->gravity=Vector3r(0,0,-9.81);
			rootBody->engines.push_back(ge);
		//@
		shared_ptr<PhysicalActionDamper> dampingDispatcher(new PhysicalActionDamper);
			shared_ptr<CundallNonViscousForceDamping> forceDamper(new CundallNonViscousForceDamping);
			forceDamper->damping=0.2;
			dampingDispatcher->add(forceDamper);
			shared_ptr<CundallNonViscousMomentumDamping> momentumDamper(new CundallNonViscousMomentumDamping);
			momentumDamper->damping=0.2;
			dampingDispatcher->add(momentumDamper);
			rootBody->engines.push_back(dampingDispatcher);
		//@
		shared_ptr<PhysicalActionApplier> applyActionDispatcher(new PhysicalActionApplier);
			applyActionDispatcher->add(new NewtonsForceLaw);
			applyActionDispatcher->add(new NewtonsMomentumLaw);
			rootBody->engines.push_back(applyActionDispatcher);
		//@
		shared_ptr<StateMetaEngine> positionIntegrator(new StateMetaEngine);
			positionIntegrator->add(new LeapFrogPositionIntegrator);
			rootBody->engines.push_back(positionIntegrator);
		//@
		shared_ptr<StateMetaEngine> orientationIntegrator(new StateMetaEngine);
			orientationIntegrator->add(new LeapFrogOrientationIntegrator);
			rootBody->engines.push_back(orientationIntegrator);
	//@
	// set default values for Shop	
	Shop::setDefault("body_sdecGroupMask",1);
	Shop::setDefault("phys_density",2400.); Shop::setDefault("phys_young",30e9); Shop::setDefault("phys_poisson",.3);
	Shop::setDefault("aabb_randomColor",false);Shop::setDefault("shape_randomColor",false); Shop::setDefault("mold_randomColor",false);
	Shop::setDefault("aabb_color",Vector3r(0,1,0)); Shop::setDefault("shape_color",Vector3r(1,0,0)); Shop::setDefault("mold_color",Vector3r(1,0,0));

	//@
	shared_ptr<Body> box=Shop::box(Vector3r(0,0,0),Vector3r(.5,.5,.5));
	box->isDynamic=false;
	rootBody->bodies->insert(box);
	
	//@
	if(false){
		Vector3r extents(.5,.5,.5);
		shared_ptr<Body> b=shared_ptr<Body>(new Body(body_id_t(0),0));
		b->isDynamic=true;
		
		// phys
		shared_ptr<BodyMacroParameters> physics(new BodyMacroParameters);
		physics->mass=8*extents[0]*extents[1]*extents[2]*2400;
		physics->inertia=Vector3r(physics->mass*(4*extents[1]*extents[1]+4*extents[2]*extents[2])/12.,physics->mass*(4*extents[0]*extents[0]+4*extents[2]*extents[2])/12.,physics->mass*(4*extents[0]*extents[0]+4*extents[1]*extents[1])/12.);
		physics->se3=Se3r(Vector3r(0,0,0),Quaternionr::Identity());
		physics->young=30e9;
		physics->poisson=.3;
		b->physicalParameters=physics;

		// aabb
		shared_ptr<Aabb> aabb(new Aabb);
		aabb->diffuseColor=Vector3r(0,1,0);
		b->bound=aabb;

		// mold
		shared_ptr<Box> mold(new Box);
		mold->extents=extents;
		mold->diffuseColor=Vector3r(1,0,0);
		b->shape=mold;

		rootBody->bodies->insert(b);
	}

	//@
	Shop::setDefault("shape_color",Vector3r(0,1,0)); Shop::setDefault("mold_color",Vector3r(0,1,0));
	shared_ptr<Body> sphere(Shop::sphere(Vector3r(0,0,2),1));
	rootBody->bodies->insert(sphere);
	
	//@
	rootBody->dt=.2*Shop::PWaveTimeStep(rootBody);

	//@
	return true;
}

YADE_REQUIRE_FEATURE(PHYSPAR);

